File: pandoc-runner/lib/pandoc_converter.rb
# frozen_string_literal: true

require_relative 'command_executor'

# PandocConverterクラス
# pandocコマンドを実行してmarkdownからredmine-textileに変換する
class PandocConverter
  class ConversionError < StandardError; end

  def initialize(command_executor: CommandExecutor.new)
    @command_executor = command_executor
    validate_pandoc_availability
  end

  # markdownからredmine-textileに変換する
  # @param markdown_text [String] 変換元のmarkdownテキスト
  # @return [String] 変換後のredmine-textileテキスト
  # @raise [ConversionError] 変換に失敗した場合
  def convert_markdown_to_redmine_textile(markdown_text)
    raise ArgumentError, 'markdown_text must be a string' unless markdown_text.is_a?(String)

    stdout, stderr, status = @command_executor.execute(
      'pandoc', '-f', 'markdown', '-t', 'textile',
      stdin_data: markdown_text
    )

    raise ConversionError, "Pandoc conversion failed: #{stderr}" unless status.success?

    stdout
  rescue Errno::ENOENT
    raise ConversionError, 'pandoc command not found'
  end

  private

  # pandocコマンドが利用可能かチェック
  def validate_pandoc_availability
    _, _, status = @command_executor.execute('pandoc', '--version')
    raise ConversionError, 'pandoc command not available' unless status.success?
  rescue Errno::ENOENT
    raise ConversionError, 'pandoc command not found'
  end
end

File: pandoc-runner/src/PandocConverter.ts
import { CommandExecutor, FilterSelector, ConversionError } from './types';
import { DefaultCommandExecutor } from './CommandExecutor';

/**
 * デフォルトのフィルターセレクター
 */
export class DefaultFilterSelector implements FilterSelector {
  /**
   * フィルターパスを取得する
   * @param fromFormat 変換元フォーマット
   * @param toFormat 変換先フォーマット
   * @returns フィルターパス（なければnull）
   */
  getFilterPath(fromFormat: string, toFormat: string): string | null {
    if (fromFormat === 'markdown' && toFormat === 'redmine-textile') {
      // TODO: adjust path as needed - use relative path from project root
      return '../dist/src/ToRedmine.js';
    }
    return null;
  }
}

/**
 * pandocコマンドを実行してmarkdownからredmine-textileに変換するクラス
 */
export class PandocConverter {
  private commandExecutor: CommandExecutor;
  private filterSelector: FilterSelector;

  constructor(
    commandExecutor: CommandExecutor = new DefaultCommandExecutor(),
    filterSelector: FilterSelector = new DefaultFilterSelector()
  ) {
    this.commandExecutor = commandExecutor;
    this.filterSelector = filterSelector;
  }

  /**
   * pandocコマンドが利用可能かチェック
   */
  async validatePandocAvailability(): Promise<void> {
    try {
      const result = await this.commandExecutor.execute('pandoc', ['--version']);
      if (result.exitCode !== 0) {
        throw new ConversionError('pandoc command not available');
      }
    } catch (error) {
      if (error instanceof Error && error.message.includes('ENOENT')) {
        throw new ConversionError('pandoc command not found');
      }
      throw new ConversionError('pandoc command not available');
    }
  }

  /**
   * markdownからredmine-textileに変換する
   * @param markdownText 変換元のmarkdownテキスト
   * @returns 変換後のredmine-textileテキスト
   */
  async convertMarkdownToRedmineTextile(markdownText: string): Promise<string> {
    if (typeof markdownText !== 'string') {
      throw new Error('markdown_text must be a string');
    }

    const filterPath = this.filterSelector.getFilterPath('markdown', 'redmine-textile');
    const commandArgs = ['pandoc', '-f', 'markdown', '-t', 'textile'];
    
    if (filterPath) {
      commandArgs.push('--filter', filterPath);
    }

    try {
      const result = await this.commandExecutor.execute(
        commandArgs[0],
        commandArgs.slice(1),
        { stdin: markdownText }
      );

      if (result.exitCode !== 0) {
        throw new ConversionError(`Pandoc conversion failed: ${result.stderr}`);
      }

      return result.stdout;
    } catch (error) {
      if (error instanceof Error && error.message.includes('ENOENT')) {
        throw new ConversionError('pandoc command not found');
      }
      if (error instanceof ConversionError) {
        throw error;
      }
      throw new ConversionError(`Conversion failed: ${error}`);
    }
  }
}
